client/context/query.go

```diff
@@ -0,0 +1,309 @@
+	account, err := ctx.GetAccount(address)
+	if err != nil {
+		return 0, err
+	}
+
+	return account.GetSequence(), nil
+}
+
+// BroadcastTx broadcasts transaction bytes to a Tendermint node.
+func (ctx QueryContext) BroadcastTx(tx []byte) (*ctypes.ResultBroadcastTxCommit, error) {
+	node, err := ctx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := node.BroadcastTxCommit(tx)
+	if err != nil {
+		return res, err
+	}
+
+	if res.CheckTx.Code != uint32(0) {
+		return res, errors.Errorf("checkTx failed: (%d) %s",
+			res.CheckTx.Code,
+			res.CheckTx.Log)
+	}
+
+	if res.DeliverTx.Code != uint32(0) {
+		return res, errors.Errorf("deliverTx failed: (%d) %s",
+			res.DeliverTx.Code,
+			res.DeliverTx.Log)
+	}
+
+	return res, err
+}
+
+// BroadcastTxAsync broadcasts transaction bytes to a Tendermint node
+// asynchronously.
+func (ctx QueryContext) BroadcastTxAsync(tx []byte) (*ctypes.ResultBroadcastTx, error) {
+	node, err := ctx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := node.BroadcastTxAsync(tx)
+	if err != nil {
+		return res, err
+	}
+
+	return res, err
+}
+
+// EnsureAccountExists ensures that an account exists for a given context. An
+// error is returned if it does not.
+func (ctx QueryContext) EnsureAccountExists() error {
+	addr, err := ctx.GetFromAddress()
+	if err != nil {
+		return err
+	}
+
+	accountBytes, err := ctx.QueryStore(auth.AddressStoreKey(addr), ctx.AccountStore)
+	if err != nil {
+		return err
+	}
+
+	if len(accountBytes) == 0 {
+		return ErrInvalidAccount(addr)
+	}
+
+	return nil
+}
+
+// EnsureAccountExistsFromAddr ensures that an account exists for a given
+// address. Instead of using the context's from name, a direct address is
+// given. An error is returned if it does not.
+func (ctx QueryContext) EnsureAccountExistsFromAddr(addr sdk.AccAddress) error {
+	accountBytes, err := ctx.QueryStore(auth.AddressStoreKey(addr), ctx.AccountStore)
+	if err != nil {
+		return err
+	}
+
+	if len(accountBytes) == 0 {
+		return ErrInvalidAccount(addr)
+	}
+
+	return nil
+}
+
+// EnsureBroadcastTx broadcasts a transactions either synchronously or
+// asynchronously based on the context parameters. The result of the broadcast
+// is parsed into an intermediate structure which is logged if the context has
+// a logger defined.
+func (ctx QueryContext) EnsureBroadcastTx(txBytes []byte) error {
+	if ctx.Async {
+		return ctx.ensureBroadcastTxAsync(txBytes)
+	}
+
+	return ctx.ensureBroadcastTx(txBytes)
+}
+
+func (ctx QueryContext) ensureBroadcastTxAsync(txBytes []byte) error {
+	res, err := ctx.BroadcastTxAsync(txBytes)
+	if err != nil {
+		return err
+	}
+
+	if ctx.JSON {
+		type toJSON struct {
+			TxHash string
+		}
+
+		if ctx.Logger != nil {
+			resJSON := toJSON{res.Hash.String()}
+			bz, err := ctx.Codec.MarshalJSON(resJSON)
+			if err != nil {
+				return err
+			}
+
+			ctx.Logger.Write(bz)
+			io.WriteString(ctx.Logger, "\n")
```

Comment:
Would it be better to use the logger here instead of just `io.WriteString` and `fmt.Printf` in various places? Although I guess we want machine-readable output and the logger adds extra info - maybe we could have a `json` logger and a `human-readable` logger though.

Reply:
Hmmm not sure I follow. The `io.WriteString` is just a convenience method for writing a newline. I can easily just append a newline byte to `bz`. The actual logger here is `os.Stdout`. I also don't see any `fmt.Printf`s.